File: LinkedLists/C++/Linkedlist.cpp
#include<iostream>
#include<stdlib.h>
#include<malloc.h>

#include "Node.h"

using std::cout;
using std::cin;
using std::end;

template <typename T>
class Linkedlist{

    public:
            Node<T> *head;

            Linkedlist(): head(nullptr){}

            void printLinkedlist(const Linkedlist& l){
                if(l.head == nullptr)
                cout << "The list is empty \n";
                else{
                    Node<T> *aux = l.head;
                    while(aux != nullptr){
                        cout << aux->data << " ";
                        aux = aux->next;
                    }
                    cout << "\n";
                }
            }

            void insertLinkedlist(Linkedlist& l,T d){
                if(l.head == nullptr)
                l.head = new Node<T>(d);
                else{
                    Node<T> *aux = l.head;
                    while(aux->next != nullptr)
                    aux = aux->next;
                    
                    aux->next = new Node<T>(d);
                }
            }

            void deleteLinkedlist(Linkedlist& l,T d){
                if(l.head == nullptr)
                cout << "the list is empty \n";
                else{
                    Node<T> *aux = l.head;
                    Node<T> *prev = nullptr;

                    while(aux != nullptr && aux->data != d){
                        prev = aux;
                        aux = aux->next;
                    }

                    if(aux != nullptr){
                        if(prev == nullptr)
                        l.head = l.head->next;
                        else prev->next = aux->next;
                    }
                }
            }

            Node<T>* searchLinkedlist(const Linkedlist& l,T d){
                
                Node<T> *res = nullptr;
                
                if(l.head == nullptr)
                cout << "The list is empty \n";
                else{
                    Node<T> *aux = l.head;
                    while(aux != nullptr && aux->data != d)
                    aux = aux->next;

                    if(aux != nullptr)
                    res = new Node<T>(d);
                }
                
                return res;
            }

            void reverseLinkedlist(Linkedlist<T>& l){


                if(l.head != nullptr && l.head->next != nullptr){
                    Node<T> *prev = nullptr;
                    Node<T> *curr = l.head;
                    Node<T> *succ = l.head->next;

                    while(succ != nullptr){
                        curr->next = prev;
                        prev = curr;

                        curr = succ;
                        succ = succ->next;
                    }
                    curr->next = prev;
                    l.head = curr;

                 
                } 

            }


            Linkedlist<T> merge(Linkedlist<T> l1, Linkedlist<T> l2){

                if(l1.head == nullptr)
                return l2;
                
                if(l2.head == nullptr)
                return l1;

                Linkedlist<T> res;

                Node<T> *aux1 = l1.head;
                Node<T> *aux2 = l2.head;
                Node<T> *temp;

                while(aux1 != nullptr && aux2 != nullptr){
                    if(aux1->data <= aux2->data){
                        if(res.head == nullptr){       
                        res.head = aux1;
                        temp = res.head;
                        aux1 = aux1->next;
                    }else{
                        temp->next = aux1;
                        aux1 = aux1->next;
                        temp = temp->next;
                    }

                }else{
                      if(res.head == nullptr){
                        res.head = aux2;
                        temp = res.head;
                        aux2 = aux2->next;
                      }else{
                            temp->next = aux2;
                            aux2 = aux2->next;
                            temp = temp->next;
                      }
                    
                    }

               }

           
               if(aux1 != nullptr)
               temp->next = aux1;
               
               if(aux2 != nullptr)
               temp->next = aux2;

                return res;
            }


        Node<T>* middleLinkedlist(Linkedlist<T> l){

            Node<T> *prev = l.head;
            Node<T> *aux = l.head;

            while(aux != nullptr){
                aux = aux->next;
                
                if(aux != nullptr){
                    prev = prev->next;
                    aux = aux->next;
                }
            }
            /*
            while(aux != nullptr){[2,1]
             aux = aux->next;
             if(aux != nullptr)
             aux = aux->next;

            if(aux != nullptr)
            prev = prev->next;
             
            }*/

            return prev;
        }


        Linkedlist<T> mergeSortLinkedlist(Linkedlist<T> l,Node<T> *ini,Node<T> *fin){

            if(ini == fin){
                //cout << ini->data << "\n";
                //cout << fin->data << "\n";
                 cout << "test 1\n";
                Linkedlist<T> temp;
                temp.head = ini;
                return temp;
            }else{
                Node<T> *mid = middleLinkedlist(l);
                //cout << mid->data << "\n";
                cout << "test \n";
                Linkedlist<T> split;
                split.head = mid->next;  
                mid->next = nullptr;

                return merge(mergeSortLinkedlist(l,ini,mid),mergeSortLinkedlist(split,split.head,fin));  
                
            }
        }
            
};

    int main(){

        int N1;
        int N2;
        Linkedlist<int> l1;
        Linkedlist<int> l2;
        
        cout << "Introduce the number of elements you would like to insert for the first list:\n";
        cin >> N1;
        
        //cout << "Introduce the number of elements you would like to insert for the second list:\n";
        //cin >> N2;
        
        int d;
        for(int i=0;i<N1;i++){
            cout << "for the first list Give me data number "<< i+1 << " \n";
            cin >> d;
            l1.insertLinkedlist(l1,d);
        }
        cout << "\n";

        /*
        for(int x=0;x<N2;x++){
            cout << "for the second list Give me data number "<< x+1 << " \n";
            cin >> d;
            l2.insertLinkedlist(l2,d);
        }
        cout << "\n";*/

        

        l1.printLinkedlist(l1);
        cout << "\n";

        /*l2.printLinkedlist(l2);
        cout << "\n";
        cout << "\n";*/

        /*Linkedlist<int> r = l1.merge(l1,l2);
        cout << "Your merged list is: \n";
        r.printLinkedlist(r);*/

        /*
        Node<int> *r = l1.middleLinkedlist(l1);
        cout << "your middle point is: " << r->data;
        cout << "\n";*/
        

        /* 4 2 0 1 5 0 19 11 1 9 [0,1,2,0,0,2,2,1]*/

        
        cout << "\n";
        Linkedlist<int> res = l1.mergeSortLinkedlist(l1,l1.head,nullptr); 
        cout << "\n";

        res.printLinkedlist(res);
        cout << "\n";
        exit(0);
    }

File: LinkedLists/Java/Linkedlist.java
import java.io.*;
import java.util.Scanner;

class Linkedlist<T>{

    Node<T> head;

    public Linkedlist(){
       this.head = null;
        
    }

    public static <T>void printList(Linkedlist<T> L){
        
        if(L.head == null)
        System.out.println("The list is empty");
        else{  
            Node<T> aux = L.head;

            while(aux != null){
                System.out.print(aux.data+" ");
                aux = aux.next;
            }
            System.out.println("");
        }
    } 


    public static <T>void insertLast(Linkedlist<T> L, T n){

        Node<T> aux = new Node<>(n);

        if(L.head == null)
        L.head = aux;
        else{
            Node<T> temp = L.head;
            
            while(temp.next != null )
            temp = temp.next;

            temp.next = aux;
        }

    }

    public static <T>void insertFirst(Linkedlist<T> L, T n){
        
        Node<T> aux = new Node<>(n);

        if(L.head == null)
        L.head = aux;
        else{
            aux.next = L.head;
            L.head = aux;
        }
    }

    public static <T>void deleteElement(Linkedlist<T> L, T n){
        
            Node<T> temp = L.head;
            Node<T> prev = null;
            boolean b = false;

            while(temp != null && !b){
                if(temp.data == n)
                b = true;
            else { 
                    prev = temp;
                    temp = temp.next;
                }
            }

            if(L.head == null)
            System.out.println("The element is not found because your list is empty");
            else{
                 if(b){
                       if(prev == null)
                        L.head = L.head.next;
                        else prev.next = temp.next;
            }else System.out.println("The element is not found");   
        }
    
    }
   
   
    public static <T extends Comparable<T>>boolean searchElement(Linkedlist<T> L, T n){

        boolean b = false;
        Node<T> aux = L.head;

        while(aux != null && !b){
            if(aux.data.compareTo(n) == 0)
            b = true;
            else aux = aux.next;
        }

    return b;
        
    }

    /*
    

    public static Linkedlist insertSort(Linkedlist L){
        Linkedlist temp = null;

        return temp;
    } 

   */


   
   public static <T extends Comparable<T>>Linkedlist<T> merge(Linkedlist<T> Lone, Linkedlist<T> Ltwo){
        
        if(Lone.head == null)
        return Ltwo;

        if(Ltwo.head == null)
        return Lone;

        Linkedlist<T> result = new Linkedlist<>();
        Node<T> none = Lone.head;
        Node<T> ntwo = Ltwo.head;
        Node<T> aux = null;
        boolean b = false;

        while(none != null && ntwo != null){

            if((none.data).compareTo(ntwo.data)<=0){
                if(result.head == null){
                    result.head = none;
                    aux = result.head;
                }else{
                    aux.next = none;
                    aux = aux.next;
                }

                none = none.next;

            }else{
                    if(result.head == null){
                        result.head = ntwo;
                        aux = result.head;
                    }else{
                        aux.next = ntwo;
                        aux = aux.next;
                    }

                    ntwo = ntwo.next;
            }
           
        }
       
        if(none != null)
        aux.next = none;

        if(ntwo != null)
        aux.next = ntwo;
        
        return result;
    }



   public static <T extends Comparable<T>> Node<T> splitList(Node<T> ini, Node<T> fin){

    Node<T> auxf = ini;
    Node<T> auxb = ini;

    while(auxf != fin){
        
        auxf = auxf.next;
        if(auxf != fin)
        auxf = auxf.next;

        if(auxf != fin)
        auxb = auxb.next;
    }
    
    return auxb;

   }

   public static<T> Linkedlist<T> reverseList(Linkedlist<T> l){

    if(l.head == null || l.head.next == null)
    return l;
    else{
        Node<T> prev = null;
        Node<T> aux = l.head;
        Node<T> succ = l.head.next;

        while(succ != null){

            if(prev == null){
                prev = aux;
                aux = succ;
                succ = succ.next;

                prev.next = null;
                aux.next = prev;
            }else{
                prev = aux;
                aux = succ;
                succ = succ.next;

                aux.next = prev;
            }
        }

        l.head = aux;
    }

    return l;
   }


public static<T extends Comparable<T>> Linkedlist<T> mergeSortList(Node<T> ini, Node<T> fin){  

    if(ini == fin){
        Linkedlist<T> aux = new Linkedlist<>();
        aux.head = ini;
        return aux;
    }else{
        Node<T> split = splitList(ini,fin);
        Node<T> mid = split.next;
        split.next = null;
        return merge(mergeSortList(ini,split),mergeSortList(mid,fin));
    }

   }

    public static void main(String args[]){


        Scanner sc = new Scanner(System.in);

        int N1;
        //int N2;

        Linkedlist<Integer> mylist1 = new Linkedlist<>();
        //Linkedlist<Integer> mylist2 = new Linkedlist<>();

        System.out.println("Give me the size of your first list:");
        N1 = sc.nextInt();

      

        System.out.println("Give me the elements for your first list: \n");
        int i = 0;
        int e = 0;
        while(i < N1){
            e = sc.nextInt();
            insertLast(mylist1,e);
            i++;
        }

        System.out.println("Your list is:");
        printList(mylist1);
        System.out.println("");

        /*
        System.out.println("Give me the size of your second list:");
        N2 = sc.nextInt();

        System.out.println("Give me the elemenets for your second list: \n");
        i = 0;
        e = 0;
        while(i < N2){
            e = sc.nextInt();
            insertLast(mylist2,e);
            i++;
        }*/

        
        Node<Integer> t = splitList(mylist1.head,null);
        System.out.println("mid "+t.data);
        
     
       // System.out.println("Your reversed list is:");
        //Linkedlist<Integer> result = reverseList(mylist1);

        //Linkedlist<Integer> result = mergeSortList(mylist1.head,null); /* 4 2 0 1 5 0 19 11 1 9 [0,1,2,0,0,2,2,1]*/

        //printList(result);
        System.out.println("");
        

    }
}